### ISSUE TITLE
Window frame glitches

### ISSUE THREAD

**Issue by TAforever** [riverwm/river#807]
Seen with imv and telegram programs. Telegram starts at system startup and when I send it to another tag, the window frame remembers its previous size until you click the mouse or shift the focus from the keyboard.
![glitch](https://user-images.githubusercontent.com/52626363/225736262-c54cbc3e-25e2-4de3-864a-e1dbfd42143b.png)

**Comment by Leon-Plickat** [comment]
What river version are you using?

**Comment by ifreund** [comment]
For imv this is probably https://todo.sr.ht/~exec64/imv/47.

**Comment by TAforever** [comment]
> What river version are you using?

Yesterday I built river-git with the latest commits. If you need I can provide a config, but there is nothing special there, just set `riverctl rule-add ssd -app-id '*'` because now by default all gtk applications use csd.

**Comment by TAforever** [comment]
> For imv this is probably https://todo.sr.ht/~exec64/imv/47.

I have noticed that this happens with almost all non-gtk applications.

**Comment by TAforever** [comment]
> I can only reproduce it with imv, what others app did you saw that?

telegram, retroarch, qt5ct, crow-translate.

**Comment by novakne** [comment]
But is that with imv beside these programs? Because I quickly test qt5ct and retroarch and I still can't reproduce it

**Comment by TAforever** [comment]
> But is that with imv beside these programs? Because I quickly test qt5ct and retroarch and I still can't reproduce it

No.

**Comment by tleydxdy** [comment]
seeing similar things with brave browser, the borders will be the wrong size sometimes when it launches (the size seems the be determined by the last float size)

**Comment by MaxVerevkin** [comment]
I can reproduce this with my client. River behaves weirdly if a client schedules a frame and commits the surface in response to configure event, instead of rendering a new frame right away.

**Comment by vainiovano** [comment]
I can reproduce this with speedcrunch (a Qt5 app) and qt6ct, but it occurs pretty rarely. Normally moving the window to an empty tag looks like this with WAYLAND_DEBUG=1:
```
[1865839.266] xdg_toplevel@23.configure(1916, 1051, array[4])
[1865839.285] xdg_surface@22.configure(77111)
[1865839.474]  -> xdg_surface@22.set_window_geometry(0, 0, 1916, 1051)
[1865839.504]  -> xdg_surface@22.ack_configure(77111)
[1865839.551]  -> wl_surface@19.frame(new id wl_callback@27)
[1865839.573]  -> wl_surface@19.attach(wl_buffer@37, 0, 0)
[1865839.590]  -> wl_surface@19.damage_buffer(0, 0, 1916, 1051)
[1865839.607]  -> wl_surface@19.commit()
```
but when the bug occurs, it looks like this instead:
```
[1872023.448] xdg_toplevel@23.configure(1916, 1051, array[8])
[1872023.457] xdg_surface@22.configure(77197)
[1872023.528]  -> xdg_surface@22.set_window_geometry(0, 0, 1916, 1051)
[1872023.542]  -> wl_compositor@8.create_region(new id wl_region@33)
[1872023.550]  -> wl_region@33.add(0, 0, 1916, 1051)
[1872023.559]  -> wl_surface@19.set_opaque_region(wl_region@33)
[1872023.565]  -> wl_region@33.destroy()
[1872023.573]  -> xdg_surface@22.ack_configure(77197)
[1872023.596]  -> wl_surface@19.frame(new id wl_callback@36)
[1872023.606]  -> wl_surface@19.attach(wl_buffer@32, 0, 0)
[1872023.614]  -> wl_surface@19.damage_buffer(0, 0, 1148, 1051)
[1872023.622]  -> wl_surface@19.commit()
[1872025.308] wl_display@1.delete_id(35)
[1872028.266]  -> wl_surface@19.attach(wl_buffer@37, 0, 0)
[1872028.319]  -> wl_surface@19.damage_buffer(0, 0, 1916, 1051)
[1872028.327]  -> wl_surface@19.commit()
[1872028.364] wl_callback@35.done(77198)
```

`wl_buffer@32` is the smaller buffer Qt was using when the window was still tiled with my terminal. Could this be a race condition in Qt? It does some pretty weird thread stuff in its Wayland backend...

**Comment by ifreund** [comment]
> `wl_buffer@32` is the smaller buffer Qt was using when the window was still tiled with my terminal. Could this be a race condition in Qt? It does some pretty weird thread stuff in its Wayland backend...

Yes, from the logs you shared that instance is a Qt bug very similar to the imv bug mentioned above :/

**Comment by ifreund** [comment]
Looks like I can reproduce this with a debug build of foot running under valgrind. Since I don't think foot is misbehaving terribly I suspect a race condition in river.

**Comment by ideasman42** [comment]
This is also an issue with Blender, I tried for a while but didn't manage to find a fix that worked on the Blender side (as far as I could see Blender is following the spec), although I could have missed something of course.

**Comment by zgb124** [comment]
I'm seeing this in imv and chromium when launched with `--ozone-platform=wayland`.

**Comment by Co3co** [comment]
I'm able to reproduce this with `mpv --player-operation-mode=pseudo-gui`, st Xwayland windows, and alacritty.
It is very easy to reproduce with alacritty.
1. Open 3 or more alacritty windows on an empty tag with rivertile's default tiling
2. Close windows with river's `close` command
3. Repeat until you get the bug

**Comment by charbelnicolas** [comment]
I can reproduce this with chromium and blender too

**Comment by ifreund** [comment]
Alright, looks like enough clients are buggy for me to implement a workaround + shame message for them: 95da9b5875b03198ab5d5c031f3b9e9c57843956

**Comment by TAforever** [comment]
Apparently the problem is finally fixed, but I noticed two strange things: the first is the inability to record using obs, it does not see the pipewire capture source and the second: if I open imv from kitty and there is a firefox window next to it, then I will not see the image but I will see the wallpaper where the imv window should be

**Comment by ifreund** [comment]
> the first is the inability to record using obs, it does not see the pipewire capture source

This is entirely unrelated to this issue and probably due to your xdg-desktop-portal-wlr configuration

> I open imv from kitty and there is a firefox window next to it, then I will not see the image but I will see the wallpaper where the imv window should be

Feel free to open a new issue for this, ideally with WAYLAND_DEBUG=1 logs of imv. I strongly suspect an imv bug though.

**Comment by zgb124** [comment]
I have recently updated to a build that includes this fix, and am still seeing glitches (albeit different ones than before) with Chromium. I am using #909 and #942 with stacktile as the layout generator, if that is relevant. Both imv and zathura work correctly with this fix.

This is what the glitches look like:

https://github.com/riverwm/river/assets/11862293/77f38eb3-e9c4-4c16-a120-b159e4286eda
